framework/web/DbSession.php

```diff
@@ -74,16 +74,15 @@ class DbSession extends MultiFieldSession
      */
     public $sessionTable = '{{%session}}';
 
-
     /**
      * Initializes the DbSession component.
      * This method will initialize the [[db]] property to make sure it refers to a valid DB connection.
      * @throws InvalidConfigException if [[db]] is invalid.
      */
-    public function init()
+    public function __construct(array $config = [])
     {
-        parent::init();
         $this->db = Instance::ensure($this->db, Connection::className());
+        parent::__construct($config);
```

Comment:
Do we need to do this change?

Reply:
Yes, or exception will be thrown in case of DbSession.
```bash
Error: Call to a member function getQueryBuilder() on string in /app/vendor/yiisoft/yii2/db/Query.php:129
Stack trace:
#0 /app/vendor/yiisoft/yii2/db/Query.php(276): yii\db\Query->createCommand('db')
#1 /app/vendor/yiisoft/yii2/web/DbSession.php(152): yii\db\Query->scalar('db')
#2 [internal function]: yii\web\DbSession->readSession('335a93c75caa6da...')
#3 /app/vendor/yiisoft/yii2/web/Session.php(137): session_start()
#4 /app/vendor/yiisoft/yii2/web/Session.php(938): yii\web\Session->open()
#5 /app/vendor/yiisoft/yii2/web/Session.php(494): yii\web\Session->unfreeze()
#6 /app/vendor/yiisoft/yii2/base/Component.php(174): yii\web\Session->setTimeout(1209600)
#7 /app/vendor/yiisoft/yii2/BaseYii.php(532): yii\base\Component->__set('timeout', 1209600)
#8 /app/vendor/yiisoft/yii2/base/BaseObject.php(106): yii\BaseYii::configure(Object(yii\web\DbSession), Array)
#9 [internal function]: yii\base\BaseObject->__construct(Array)
#10 /app/vendor/yiisoft/yii2/di/Container.php(381): ReflectionClass->newInstanceArgs(Array)
#11 /app/vendor/yiisoft/yii2/di/Container.php(156): yii\di\Container->build('yii\\web\\DbSessi...', Array, Array)
```

Reply:
Here's more details. https://github.com/yiisoft/yii2/pull/15523#issuecomment-359675906

Reply:
What if you specify `db` in constructor's `$config`? You just ignoring this setting.

Reply:
Not sure what do you mean.

Reply:
yes I can, but why, if there IS not breaking BC option.

Reply:
This is a breaking BC option. If I do:

```php 
$session = new DbSession(['db' => 'sessionDb']);
```

I expect that `Yii::$app->sessionDb` will be used as session DB. It worked that before, but now you just using default value (`Yii::$app->db`) during initialization, and at the end `db` is not initialized (which will give you errors like above).

Reply:
I called parent implementation at the end, so you it supposed to be overriden by the end of initialization from the array config.

Reply:
I suggest to write tests for it, I'm pretty sure that you're wrong and it will not work, since you're removed old `init()` implementation and expect to use `db` before it is configured.

Reply:
Ok, may I have exactly that test for situation you're describing, so I'm able to continue working on this?

Reply:
@rob006 do you have some time to pull request a test to this PR branch?

Reply:
@samdark No, I'm quite busy recently. But I'm quite sure that this will fail:

```php
$dbSession = new DbSession(['db' => [
    'class' => MyDbConection::class,
    // ...
]);
$this->assertInstanceOf(MyDbConection::class, $dbSession->db);
```

Reply:
@StalkAlex Could you try reverse `init()` body and check if error mentioned in https://github.com/yiisoft/yii2/pull/15523#discussion_r166183698 still occurs?

```php
    public function init()
    {
        $this->db = Instance::ensure($this->db, Connection::className());
        parent::init();
    }
```

Reply:
@samdark @rob006 I'm checking and getting the same result. I can't understand: BaseObject's constructor called as suppose to be according to stack trace, it contains `$this->init()` but `DbSession::init()` hasn't been called.

Reply:
Same result means the change isn't needed?

Reply:
It means yes change is needed. 
![bug](https://user-images.githubusercontent.com/2943154/35932506-97238840-0c59-11e8-9046-db24fff467f5.png)
It doesn't have chance to call line 109 to init $db, it happens on line 107, when it tries to set timeout and then start session due to new changes.

Reply:
Hmm, fixed by https://github.com/yiisoft/yii2/pull/15523/commits/d313bd586fbbdb0b7c716e99d11db4d3c44ca0b8?